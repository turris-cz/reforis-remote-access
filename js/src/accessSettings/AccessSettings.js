/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useEffect } from "react";
import PropTypes from "prop-types";

import { useAPIGet } from "foris";

import API_URLs from "API";
import Settings from "./Settings";

AccessSettings.propTypes = {
    ws: PropTypes.object.isRequired,
};

export default function AccessSettings({ ws }) {
    const [getAuthorityResponse, getAuthority] = useAPIGet(API_URLs.authority);
    useEffect(() => {
        getAuthority();
    }, [getAuthority]);

    const [getSettingsResponse, getSettings] = useAPIGet(API_URLs.settings);
    useEffect(() => {
        getSettings();
    }, [getSettings]);

    return (
        <>
            <h1>{_("Access Settings")}</h1>
            <p>
                {_(
                    "Here you can set up your router to be configured remotely. The remote configuration is done via secure encrypted connection and each client is required to have a token issued by this device."
                )}
            </p>
            <p>
                {_(
                    "This can be useful when you plan to access configuration interface another Turris router or you intend to write your own client."
                )}
            </p>
            <Settings
                apiState={[
                    getAuthorityResponse.state,
                    getSettingsResponse.state,
                ]}
                ws={ws}
                authority={getAuthorityResponse.data}
                settings={getSettingsResponse.data}
                onAuthoritySuccess={getAuthority}
                onSettingsSuccess={getSettings}
            />
        </>
    );
}
