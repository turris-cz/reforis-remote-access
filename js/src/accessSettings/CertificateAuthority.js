/*
 * Copyright (C) 2019-2021 CZ.NIC z.s.p.o. (https://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useEffect } from "react";
import PropTypes from "prop-types";

import {
    Button,
    useAlert,
    useAPIDelete,
    API_STATE,
    formFieldsSize,
} from "foris";

import API_URLs from "API";

export const CA_STATUS = {
    MISSING: "missing",
    READY: "ready",
    GENERATING: "generating",
};

CertificateAuthority.propTypes = {
    authority: PropTypes.object.isRequired,
    accessEnabled: PropTypes.bool.isRequired,
    onSuccess: PropTypes.func.isRequired,
};

export default function CertificateAuthority({
    authority,
    accessEnabled,
    onSuccess,
}) {
    const [setAlert] = useAlert();

    const [deleteResponse, deleteCA] = useAPIDelete(API_URLs.authority);
    useEffect(() => {
        if (deleteResponse.state === API_STATE.SUCCESS) {
            onSuccess();
        } else if (deleteResponse.state === API_STATE.ERROR) {
            setAlert(_("Cannot delete certificate authority."));
        }
    }, [deleteResponse, onSuccess, setAlert]);

    if (authority.status !== CA_STATUS.READY) {
        return null;
    }

    return (
        <div className={formFieldsSize}>
            <h2>{_("Certificate Authority")}</h2>
            {accessEnabled ? (
                <p>
                    {_(
                        "You can't delete certificate authority while remote access is enabled. To delete it, you need to disable the access first."
                    )}
                </p>
            ) : (
                <>
                    <p>
                        {_(
                            "Your certificate authority is set up properly. Please note that if you delete it, all clients will have their access revoked."
                        )}
                    </p>
                    <div className="text-right">
                        <Button
                            onClick={() => deleteCA()}
                            loading={deleteResponse.state === API_STATE.SENDING}
                            forisFormSize
                        >
                            {_("Delete Certificate Authority")}
                        </Button>
                    </div>
                </>
            )}
        </div>
    );
}
