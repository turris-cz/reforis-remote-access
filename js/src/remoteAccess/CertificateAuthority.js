/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useEffect } from "react";
import PropTypes from "prop-types";

import {
    Button, useAlert, useAPIDelete, API_STATE,
} from "foris";

import API_URLs from "API";

export const CA_STATUS = {
    MISSING: "missing",
    READY: "ready",
    GENERATING: "generating",
};

CertificateAuthority.propTypes = {
    authority: PropTypes.object.isRequired,
    accessEnabled: PropTypes.bool.isRequired,
    onSuccess: PropTypes.func.isRequired,
};

export default function CertificateAuthority({ authority, accessEnabled, onSuccess }) {
    const [setAlert] = useAlert();

    const [deleteResponse, deleteCA] = useAPIDelete(API_URLs.authority);
    useEffect(() => {
        if (deleteResponse.state === API_STATE.SUCCESS) {
            onSuccess();
        } else if (deleteResponse.state === API_STATE.ERROR) {
            setAlert(_("Cannot delete certificate authority"));
        }
    }, [deleteResponse, onSuccess, setAlert]);

    if (authority.status !== CA_STATUS.READY) {
        return null;
    }

    return (
        <>
            <h3>{_("Certificate Authority")}</h3>
            {accessEnabled
                ? <p>{_("You can't delete certificate authority while remote access is enabled. In order to delete it you need to disable the access first.")}</p>
                : (
                    <>
                        <p>{_("Your certificate authority is set up properly. Please note that if you delete it all clients will have their access revoked.")}</p>
                        <Button
                            onClick={() => deleteCA()}
                            loading={deleteResponse.state === API_STATE.SENDING}
                            forisFormSize
                        >
                            {_("Delete certificate authority")}
                        </Button>
                    </>
                )}
        </>
    );
}
