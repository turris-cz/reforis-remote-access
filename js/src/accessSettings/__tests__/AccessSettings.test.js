/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import { render, wait } from "foris/testUtils/customTestRender";
import { WebSockets } from "foris";
import { mockJSONError } from "foris/testUtils/network";
import mockAxios from 'jest-mock-axios';

import AccessSettings from "../AccessSettings";
import { CA_STATUS } from "../CertificateAuthority";

describe("<AccessSettings />", () => {
    const webSockets = new WebSockets();
    let container;

    beforeEach(() => {
        ({ container } = render(<AccessSettings ws={webSockets} />));
    });

    it("should render spinner", () => {
        expect(container).toMatchSnapshot();
    });

    it("should render error", async () => {
        expect(mockAxios.get).toBeCalledWith("/reforis/remote-access/api/authority", expect.anything());
        expect(mockAxios.get).toBeCalledWith("/reforis/remote-access/api/settings", expect.anything());
        mockJSONError();
        await wait(() => expect(container).toMatchSnapshot());
    });

    it("should handle missing certificate", async () => {
        // Response to GET authority
        mockAxios.mockResponse({ data: { status: CA_STATUS.MISSING }});
        // Response to GET settings
        mockAxios.mockResponse({ data: { enabled: false }});
        await wait(() => expect(container).toMatchSnapshot());
    });

    it("should handle generating certificate", async () => {
        // Response to GET authority
        mockAxios.mockResponse({ data: { status: CA_STATUS.GENERATING }});
        // Response to GET settings
        mockAxios.mockResponse({ data: { enabled: false }});
        await wait(() => expect(container).toMatchSnapshot());
    });

    it("should handle ready certificate", async () => {
        // Response to GET authority
        mockAxios.mockResponse({ data: { status: CA_STATUS.READY }});
        // Response to GET settings
        mockAxios.mockResponse({ data: { enabled: false }});
        await wait(() => expect(container).toMatchSnapshot());
    });
});
