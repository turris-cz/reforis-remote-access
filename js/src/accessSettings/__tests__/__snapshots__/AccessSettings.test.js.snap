// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AccessSettings /> should handle generating certificate 1`] = `
<div>
  <h1>
    Access Settings
  </h1>
  <p>
    Here you can set up your router to be configured remotely. The remote configuration is done via secure encrypted connection and each client is required to have a token issued by this device.
  </p>
  <p>
    This can be useful when you plan to access configuration interface another Turris router or you intend to write your own client.
  </p>
  <h3>
    Generating certificate authority
  </h3>
  <div
    class="spinner-wrapper my-3 text-center"
  >
    <div
      class="spinner-border"
      role="status"
    >
      <span
        class="sr-only"
      />
    </div>
  </div>
  <p>
    Your certificate authority is now being generated. It usually takes a few minutes. Settings will appear here automatically once the authority is ready.
  </p>
</div>
`;

exports[`<AccessSettings /> should handle missing certificate 1`] = `
<div>
  <h1>
    Access Settings
  </h1>
  <p>
    Here you can set up your router to be configured remotely. The remote configuration is done via secure encrypted connection and each client is required to have a token issued by this device.
  </p>
  <p>
    This can be useful when you plan to access configuration interface another Turris router or you intend to write your own client.
  </p>
  <h3>
    No certificate authority
  </h3>
  <p>
    Currently there is no certificate authority (CA) dedicated to remote access. A CA is required to generate access tokens to authenticate. To proceed you need to generate it first.
  </p>
  <button
    class="btn btn-primary  col-sm-12 col-lg-3"
    type="button"
  >
     
     
    Generate certificate authority
  </button>
</div>
`;

exports[`<AccessSettings /> should handle ready certificate 1`] = `
<div>
  <h1>
    Access Settings
  </h1>
  <p>
    Here you can set up your router to be configured remotely. The remote configuration is done via secure encrypted connection and each client is required to have a token issued by this device.
  </p>
  <p>
    This can be useful when you plan to access configuration interface another Turris router or you intend to write your own client.
  </p>
  <h3>
    Remote Access
  </h3>
  <form>
    <div
      class="form-group"
    >
      <div
        class="custom-control custom-checkbox "
      >
        <input
          class="custom-control-input"
          id="1"
          type="checkbox"
        />
        <label
          class="custom-control-label"
          for="1"
        >
          Enable remote access
        </label>
      </div>
    </div>
    <button
      class="btn btn-primary  col-sm-12 col-lg-3"
      type="submit"
    >
       
       
      Save
    </button>
  </form>
  <h3>
    Certificate Authority
  </h3>
  <p>
    Your certificate authority is set up properly. Please note that if you delete it all clients will have their access revoked.
  </p>
  <button
    class="btn btn-primary  col-sm-12 col-lg-3"
    type="button"
  >
     
     
    Delete certificate authority
  </button>
</div>
`;

exports[`<AccessSettings /> should render error 1`] = `
<div>
  <h1>
    Access Settings
  </h1>
  <p>
    Here you can set up your router to be configured remotely. The remote configuration is done via secure encrypted connection and each client is required to have a token issued by this device.
  </p>
  <p>
    This can be useful when you plan to access configuration interface another Turris router or you intend to write your own client.
  </p>
  <p
    class="text-center text-danger"
  >
    An error occurred while fetching data.
  </p>
</div>
`;

exports[`<AccessSettings /> should render spinner 1`] = `
<div>
  <h1>
    Access Settings
  </h1>
  <p>
    Here you can set up your router to be configured remotely. The remote configuration is done via secure encrypted connection and each client is required to have a token issued by this device.
  </p>
  <p>
    This can be useful when you plan to access configuration interface another Turris router or you intend to write your own client.
  </p>
  <div
    class="spinner-wrapper my-3 text-center"
  >
    <div
      class="spinner-border"
      role="status"
    >
      <span
        class="sr-only"
      />
    </div>
  </div>
</div>
`;
